docs(cn): translate reference/react/useEffect into Chinese (#1122)
reactjs/zh-hans.react.dev: src/content/reference/react/useEffect.md

}
```

> [请看下面的更多例子](#usage)。

Reviewer: ```suggestion
[请看下面的更多例子。](#usage)
```

#### 参数 {/*parameters*/}

* `setup`：处理副作用的函数。setup 函数选择性返回一个 *清理（cleanup）* 函数。在将组件首次添加到 DOM 之前，React 将运行 setup 函数。在每次依赖项变更重新渲染后，React 将首先使用旧值运行 cleanup 函数（如果你提供了该函数），然后使用新值运行 setup 函数。在组件从 DOM 中移除后，React 将最后一次运行 cleanup 函数。
 
* **可选** `dependencies`：`setup` 代码中引用的所有响应式值的列表。响应式值包括 props、state 以及所有直接在组件内部声明的变量和函数。如果你的代码检查工具 [配置了 React](/learn/editor-setup#linting)，那么它将验证每个响应式值都被正确地指定为一个依赖项。依赖项列表必须具有固定数量的项，并且必须像 `[dep1, dep2, dep3]` 这样内联编写。React 将使用 [`Object.is`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/is) 来比较每个依赖项和它先前的值。如果省略此参数，则在每次重新渲染组件之后，将重新运行副作用函数。[请参见传递依赖数组，空数组和不传递依赖项之间的区别](#examples-dependencies)。

#### 返回值 {/*returns*/}

> `useEffect` 返回 `undefined`.

Reviewer 2: ```suggestion
`useEffect` 返回 `undefined`。
```

#### 注意事项 {/*caveats*/}

* `useEffect` 是一个 Hook，因此只能在 **组件的顶层** 或自己的 Hook 中调用它。不能在循环或者条件内部调用它。如果需要，抽离出一个新组件并将状态移入其中。

* 如果你 **没有打算与某个外部系统同步**，[那么你可能不需要 Effect](/learn/you-might-not-need-an-effect)。

> * 当 StrictMode 启动时，React 将在真正的 setup 函数首次运行前，**运行一个额外的开发专有的 setup + cleanup 周期**。这是一个压力测试，确保 cleanup 逻辑“映射”到 setup 逻辑，并停止或撤消 setup 函数正在做的任何事情。如果这引起一个问题，[请实现 cleanup 函数](/learn/synchronizing-with-effects#how-to-handle-the-effect-firing-twice-in-development)。

Reviewer 3: ```suggestion
* 当严格模式启动时，React 将在真正的 setup 函数首次运行前，**运行一个额外的开发专有的 setup + cleanup 周期**。这是一个压力测试，确保 cleanup 逻辑“映射”到 setup 逻辑，并停止或撤消 setup 函数正在做的任何事情。如果这引起一个问题，[请实现 cleanup 函数](/learn/synchronizing-with-effects#how-to-handle-the-effect-firing-twice-in-development)。
```

* 如果你的一些依赖项是组件内部定义的对象或函数，则存在这样的风险，即它们将 **导致 Effect 过多地重新运行**。要解决这个问题，请删除不必要的 [对象](/reference/react/useEffect#removing-unnecessary-object-dependencies) 和 [函数](/reference/react/useEffect#removing-unnecessary-function-dependencies) 依赖项。你还可以 [抽离状态更新](/reference/react/useEffect#updating-state-based-on-previous-state-from-an-effect) 和 [非响应式的逻辑](/reference/react/useEffect#reading-the-latest-props-and-state-from-an-effect) 到 Effect 之外。

> * 如果你的 Effect 不是由交互（比如点击）引起的，那么React 会让浏览器 **在运行 Effect 前先绘制出更新后的屏幕**。如果你的 Effect 正在做一些视觉相关的事情（例如，定位一个 tooltip），并且有显著的延迟（例如，它会闪烁），那么将 `useEffect` 替换为 [`useLayoutEffect`](/reference/react/useLayoutEffect)。

Reviewer 3: ```suggestion
* 如果你的 Effect 不是由交互（比如点击）引起的，那么 React 会让浏览器 **在运行 Effect 前先绘制出更新后的屏幕**。如果你的 Effect 正在做一些视觉相关的事情（例如，定位一个 tooltip），并且有显著的延迟（例如，它会闪烁），那么将 `useEffect` 替换为 [`useLayoutEffect`](/reference/react/useLayoutEffect)。
```

}
```

你需要向 `useEffect` 传递两个参数：

1. 一个 *setup 函数* ，其 <CodeStep step={1}>setup 代码</CodeStep> 用来连接到该系统。
   - 它应该返回一个 *清理函数*（cleanup），其 <CodeStep step={2}>cleanup 代码</CodeStep> 用来与该系统断开连接。
2. 一个 <CodeStep step={3}>依赖项列表</CodeStep>，包括这些函数使用的每个组件内的值。

> **React 在必要时调用 setup 和 cleanup，这可能会发生多次**：

Reviewer 3: ```suggestion
**React 在必要时会调用 setup 和 cleanup，这可能会发生多次**：
```

> 1. 将组件添加到页面 **（挂载）** 时，将运行 <CodeStep step={1}>setup 代码</CodeStep>。

Reviewer 3: ```suggestion
1. 将组件挂载到页面时，将运行 <CodeStep step={1}>setup 代码</CodeStep>。
```

> 2. 在 <CodeStep step={3}>依赖项</CodeStep> 发生变更的组件每次重新渲染之后：

Reviewer 3: ```suggestion
2. 重新渲染 <CodeStep step={3}>依赖项</CodeStep> 变更的组件后：
```

   - 首先，使用旧的 props 和 state 运行 <CodeStep step={2}>cleanup 代码</CodeStep>。
   - 然后，使用新的 props 和 state 运行 <CodeStep step={1}>setup 代码</CodeStep>。

> 3. 在组件从页面删除 **（卸载）** 之后，<CodeStep step={2}>cleanup 代码</CodeStep> 将运行最后一次。

Reviewer 3: ```suggestion
3. 当组件从页面卸载后，<CodeStep step={2}>cleanup 代码</CodeStep> 将运行最后一次。
```

> **让我们为上面这个例子解释这个顺序**。  

Reviewer 3: ```suggestion
**用上面的代码作为例子来解释这个顺序**。  
```

> 当上面的 `ChatRoom` 组件添加到页面中时，它将使用初始 `serverUrl` 和 `roomId` 连接到聊天室。如果 `serverUrl` 或者 `roomId` 因为一个重新渲染而改变（比如说，如果用户在下拉列表中选择了一个不同的聊天室），那么你的 Effect 就会 **断开与前一个聊天室的连接，并连接到下一个聊天室**。当 `ChatRoom` 组件从页面中删除时，你的 Effect 将最后一次断开连接。

Reviewer 3: ```suggestion
当 `ChatRoom` 组件添加到页面中时，它将使用 `serverUrl` 和 `roomId` 初始值连接到聊天室。如果 `serverUrl` 或者 `roomId` 发生改变并导致重新渲染（比如用户在下拉列表中选择了一个不同的聊天室），那么 Effect 就会 **断开与前一个聊天室的连接，并连接到下一个聊天室**。当 `ChatRoom` 组件从页面中卸载时，你的 Effect 将最后一次断开连接。
```

> **为了 [帮助你发现 bug](/learn/synchronizing-with-effects#step-3-add-cleanup-if-needed)，在开发环境，React 在运行 <CodeStep step={1}>setup</CodeStep> 之前会额外运行一次<CodeStep step={1}>setup</CodeStep> 和 <CodeStep step={2}>cleanup</CodeStep>**。这是一个压力测试，用于验证 Effect 逻辑是否正确实现。如果这会导致可见的问题，那么你的 cleanup 函数就缺少一些逻辑。cleanup 函数应该停止或撤消 setup 函数正在执行的任何操作。一般来说，用户不应该能够区分只调用一次 setup（如在生产中）和 调用 *setup* → *cleanup* → *setup* 序列（如在开发中）。[查看常见解决方案](/learn/synchronizing-with-effects#how-to-handle-the-effect-firing-twice-in-development)。

Reviewer 3: ```suggestion
**为了 [帮助你发现 bug](/learn/synchronizing-with-effects#step-3-add-cleanup-if-needed)，在开发环境下，React 在运行 <CodeStep step={1}>setup</CodeStep> 之前会额外运行一次<CodeStep step={1}>setup</CodeStep> 和 <CodeStep step={2}>cleanup</CodeStep>**。这是一个压力测试，用于验证 Effect 逻辑是否正确实现。如果这会导致可见的问题，那么你的 cleanup 函数就缺少一些逻辑。cleanup 函数应该停止或撤消 setup 函数正在执行的任何操作。一般来说，用户不应该能够区分只调用一次 setup（在生产环境中）与调用 *setup* → *cleanup* → *setup* 序列（在开发环境中）。[查看常见解决方案](/learn/synchronizing-with-effects#how-to-handle-the-effect-firing-twice-in-development)。
```

> **尽量 [将每个 Effect 作为一个独立的过程编写](/learn/lifecycle-of-reactive-effects#each-effect-represents-a-separate-synchronization-process)，并且 [每次只考虑一个单独的 setup/cleanup 周期](/learn/lifecycle-of-reactive-effects#thinking-from-the-effects-perspective)**。组件是否正在挂载、更新或卸载并不重要。当你的 cleanup 逻辑正确地 “映射” 到 setup 逻辑时，你的 Effect 是可复原的，因此可以根据需要多次运行 setup 和 cleanup。

Reviewer 3: ```suggestion
**尽量 [将每个 Effect 作为一个独立的过程编写](/learn/lifecycle-of-reactive-effects#each-effect-represents-a-separate-synchronization-process)，并且 [每次只考虑一个单独的 setup/cleanup 周期](/learn/lifecycle-of-reactive-effects#thinking-from-the-effects-perspective)**。组件是否正在挂载、更新或卸载并不重要。当你的 cleanup 逻辑正确地“映射”到 setup 逻辑时，你的 Effect 是可复原的，因此可以根据需要多次运行 setup 和 cleanup 函数。
```

<Note>

Effect 可以让你的组件与某些外部系统（比如聊天服务）[保持同步](/learn/synchronizing-with-effects)。在这里，外部系统是指任何不受 React 控制的代码，例如：

* 由 <CodeStep step={1}>[`setInterval()`](https://developer.mozilla.org/en-US/docs/Web/API/setInterval)</CodeStep> 和 <CodeStep step={2}>[`clearInterval()`](https://developer.mozilla.org/en-US/docs/Web/API/clearInterval)</CodeStep> 管理的定时器。
* 使用 <CodeStep step={1}>[`window.addEventListener()`](https://developer.mozilla.org/en-US/docs/Web/API/EventTarget/addEventListener)</CodeStep> 和 <CodeStep step={2}>[`window.removeEventListener()`](https://developer.mozilla.org/en-US/docs/Web/API/EventTarget/removeEventListener)</CodeStep> 的事件订阅。
* 一个第三方的动画库，它有一个类似 <CodeStep step={1}>`animation.start()`</CodeStep> 和 <CodeStep step={2}>`animation.reset()`</CodeStep> 的 API。

**如果你没有连接到任何外部系统，[你或许不需要 Effect](/learn/you-might-not-need-an-effect)**。

</Note>

<Recipes titleText="连接到外部系统的示例" titleId="examples-connecting">

#### 连接到聊天服务器 {/*connecting-to-a-chat-server*/}

> 在这个示例中，`ChatRoom` 组件使用一个 Effect 来保持与 `chat.js` 中定义的外部系统的连接。点击“打开聊天”以显示 `ChatRoom` 组件。这个沙盒在开发模式下运行，因此有一个额外的 连接—断开 周期，就像 [这里描述的](/learn/synchronizing-with-effects#step-3-add-cleanup-if-needed)。尝试使用下拉菜单和输入框更改 `roomId` 和 `serverUrl`，并查看 Effect 如何重新连接到聊天。点击“关闭聊天”可以看到 Effect 最后一次断开连接。

Reviewer 3: ```suggestion
在这个示例中，`ChatRoom` 组件使用一个 Effect 来保持与 `chat.js` 中定义的外部系统的连接。点击“打开聊天”以显示 `ChatRoom` 组件。这个沙盒在开发模式下运行，因此有一个额外的“连接并断开”的周期，就像 [这里描述的](/learn/synchronizing-with-effects#step-3-add-cleanup-if-needed) 一样。尝试使用下拉菜单和输入框更改 `roomId` 和 `serverUrl`，并查看 Effect 如何重新连接到聊天。点击“关闭聊天”可以看到 Effect 最后一次断开连接。
```

<Solution />

#### 监听全局的浏览器事件 {/*listening-to-a-global-browser-event*/}

> 在这个例子中，外部系统就是浏览器 DOM 本身。通常，你会使用 JSX 指定事件监听，但是你不能以这种方式监听全局的 [`window`](https://developer.mozilla.org/en-US/docs/Web/API/Window) 对象。通过 Effect 可以使你连接到 `window` 对象并监听其事件。监听 `pointermove` 事件可以让你追踪光标（或手指）的位置，并更新红点以随之移动。

Reviewer 3: ```suggestion
在这个例子中，外部系统就是浏览器 DOM 本身。通常，你会使用 JSX 指定事件监听，但是你不能以这种方式监听全局的 [`window`](https://developer.mozilla.org/en-US/docs/Web/API/Window) 对象。你可以通过 Effect 连接到 `window` 对象并监听其事件。如监听 `pointermove` 事件可以让你追踪光标（或手指）的位置，并更新红点以随之移动。
```

<Solution />

#### 控制模态对话框 {/*controlling-a-modal-dialog*/}

> 在这个例子中，外部系统是浏览器 DOM。`ModalDialog` 组件渲染一个 [`<dialog>`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/dialog) 元素。它使用一个 Effect 将 `isOpen` prop 同步到 [`showModal()`](https://developer.mozilla.org/en-US/docs/Web/API/HTMLDialogElement/showModal) 和 [`close()`](https://developer.mozilla.org/en-US/docs/Web/API/HTMLDialogElement/close) 方法调用。

Reviewer 3: ```suggestion
在这个例子中，外部系统是浏览器 DOM。`ModalDialog` 组件渲染一个 [`<dialog>`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/dialog) 元素。它使用 Effect 将 `isOpen` prop 同步到 [`showModal()`](https://developer.mozilla.org/en-US/docs/Web/API/HTMLDialogElement/showModal) 和 [`close()`](https://developer.mozilla.org/en-US/docs/Web/API/HTMLDialogElement/close) 方法调用。
```

<Solution />

#### 跟踪元素可见性 {/*tracking-element-visibility*/}

> 在这个例子中，外部系统仍然是浏览器 DOM。`App` 组件展示一个长列表，然后是 `Box` 组件，然后是另一个长列表。向下滚动列表。请注意，当 `Box` 组件出现在视口中时，背景色会变成黑色。为了实现这一点，`Box` 组件使用一个 Effect 来管理一个 [`IntersectionObserver`](https://developer.mozilla.org/en-US/docs/Web/API/Intersection_Observer_API)。这个浏览器 API 会在视口中出现该 DOM 元素时通知你。

Reviewer 3: ```suggestion
在这个例子中，外部系统仍然是浏览器 DOM。`App` 组件展示一个长列表，然后是 `Box` 组件，然后是另一个长列表。试试向下滚动列表。请注意，当 `Box` 组件出现在视野中时，背景色会变成黑色。为了实现这一点，`Box` 组件使用 Effect 来管理 [`IntersectionObserver`](https://developer.mozilla.org/en-US/docs/Web/API/Intersection_Observer_API)。这个浏览器 API 会在视野中出现指定 DOM 元素时通知你。
```

---

### 控制非 React 小部件 {/*controlling-a-non-react-widget*/}

> 有时，你希望外部系统与你组件的某些 prop 或 state 保持同步。

Reviewer 3: ```suggestion
有时，你希望外部系统与你组件的某些 props 或 state 保持同步。
```

> 例如，如果你有一个没有使用 React 编写的第三方地图小部件或视频播放器组件，你可以使用一个 Effect 调用该组件上的方法，使其状态与 React 组件的当前状态相匹配。此 Effect 创建了在 `map-widget.js` 中定义的 `MapWidget` 类的实例。当你更改 `Map` 组件的 `zoomLevel` prop 时，Effect 调用类实例上的 `setZoom()` 来保持同步：

Reviewer 3: ```suggestion
例如，如果你有一个没有使用 React 编写的第三方地图小部件或视频播放器组件，你可以使用 Effect 调用该组件上的方法，使其状态与 React 组件的当前状态相匹配。此 Effect 创建了在 `map-widget.js` 中定义的 `MapWidget` 类的实例。当你更改 `Map` 组件的 `zoomLevel` prop 时，Effect 调用类实例上的 `setZoom()` 来保持同步：
```

</Sandpack>

> 直接在 Effect 中编写数据请求会显得重复，并且很难在以后添加缓存和服务端渲染等优化。[使用自定义 Hook 更简单—不管是你自己的 Hook 还是由社区维护的 Hook](/learn/reusing-logic-with-custom-hooks#when-to-use-custom-hooks)。

Reviewer 3: ```suggestion
直接在 Effect 中编写数据请求会显得重复，并且很难在以后添加缓存和服务端渲染等优化。[使用自定义 Hook 更简单——不管是你自己的 Hook 还是由社区维护的 Hook](/learn/reusing-logic-with-custom-hooks#when-to-use-custom-hooks)。
```

<DeepDive>

#### Effect 中的数据请求有什么好的替代方法 {/*what-are-good-alternatives-to-data-fetching-in-effects*/}

> 在 Effect 中编写 `fetch` 调用是 [请求数据的一种流行方式](https://www.robinwieruch.de/react-hooks-fetch-data/)，特别是在完全的客户端应用程序中。然而，这是一种非常手动的方法，而且有很大的缺点：

Reviewer 3: ```suggestion
在 Effect 中使用 `fetch` 是 [请求数据的一种流行方式](https://www.robinwieruch.de/react-hooks-fetch-data/)，特别是在完全的客户端应用程序中。然而，这是一种非常手动的方法，而且有很大的缺点：
```

> - **Effect 不在服务器上运行**。这意味着初始服务器渲染的 HTML 将只包含没有数据的 loading 状态。客户端电脑将不得不下载所有的 JavaScript 并渲染你的应用程序，仅为了发现它现在需要加载数据。这并不高效。

Reviewer 3: ```suggestion
- **Effect 不在服务器上运行**。这意味着初始服务器渲染的 HTML 将只包含没有数据的 loading 状态。客户端电脑仅为了发现它现在需要加载数据，将不得不下载所有的脚本来渲染你的应用程序。这并不高效。
```

> - **在 Effect 中直接请求数据很容易创建“网络瀑布”**。你渲染父组件，它请求一些数据，再渲染子组件, 然后它们请求它们的数据。如果网络不是很快，那么这比并行请求所有数据要慢得多。

Reviewer 3: ```suggestion
- **在 Effect 中直接请求数据很容易导致“网络瀑布”**。当你渲染父组件时，它会请求一些数据，再渲染子组件，然后重复这样的过程来请求子组件的数据。如果网络不是很快，这将比并行请求所有数据要慢得多。
```

> - **在 Effect 中直接请求数据通常意味着你不会预加载或缓存数据**。例如，如果组件卸载后重新挂载，它不得不再次请求数据。

Reviewer 3: ```suggestion
- **在 Effect 中直接请求数据通常意味着无法实现预加载或缓存数据**。例如，如果组件卸载后重新挂载，它不得不再次请求数据。
```

Author: 我认为这里表达的意思是用户“不会实现”，“无法实现”过于绝对了

> - **这不符合工效学**。 在编写 `fetch` 调用时，需要编写大量样板代码，以避免像 [竞争条件](https://maxrozen.com/race-conditions-fetching-data-react-with-useeffect) 这样的 bug。

Reviewer 3: ```suggestion
- **这不符合工效学**。 在使用 `fetch`，需要编写大量样板代码，以避免像 [竞争条件](https://maxrozen.com/race-conditions-fetching-data-react-with-useeffect) 这样的 bug。
```

> 这份缺点清单并不是针对 React 的。它适用于使用任何库在挂载时请求数据。与路由一样，要做好数据请求并非易事，因此我们推荐以下方法：

Reviewer 3: ```suggestion
这些缺点并不仅仅体现在 React 上，它可能出现在所有挂载时请求数据的地方。与路由一样，要做好数据请求并非易事，因此我们推荐以下方法：
```

- **如果使用 [框架](/learn/start-a-new-react-project#production-grade-react-frameworks)，请使用其内置的数据请求机制**。现代的 React 框架集成了高效的数据请求机制，不会受到上述问题的影响。

> 否则，请考虑使用或构建客户端缓存。流行的开源解决方案包括 [React Query](https://react-query.tanstack.com/)、[useSWR](https://swr.vercel.app/) 和 [React Router 6.4+](https://beta.reactrouter.com/en/main/start/overview)。你也可以构建自己的解决方案，在这种情况下，你可以在掌控下使用 Effect，但也要添加逻辑来处理重复的请求、缓存响应和避免“网络瀑布”（通过预加载数据或将数据需求提升到路由层面）。

Reviewer 3: ```suggestion
否则，请考虑使用或构建客户端缓存。流行的开源解决方案包括 [React Query](https://react-query.tanstack.com/)、[useSWR](https://swr.vercel.app/) 和 [React Router v6.4+](https://beta.reactrouter.com/en/main/start/overview)。你也可以构建自己的解决方案，在这种情况下，你可以在掌控下使用 Effect，但也要添加逻辑来处理重复的请求、缓存响应和避免“网络瀑布”（通过预加载数据或将数据需求提升到路由层面）。
```

    const connection = createConnection(serverUrl, roomId);
    connection.connect();
    return () => connection.disconnect();
  }, []); // 🔴 React Hook useEffect 缺少依赖项：'roomId' 和 'serverUrl'
  // ...
}
```

**要删除一个依赖项，你需要 [“证明”给代码检查工具，表明某个值 *不需要* 作为依赖项](/learn/removing-effect-dependencies#removing-unnecessary-dependencies)**。例如，你可以将 `serverUrl` 从你的组件中移除，以证明它不是响应式的，不会在重新渲染时发生变化：

```js {1,8}
const serverUrl = 'https://localhost:1234'; // 不再是一个响应式值

function ChatRoom({ roomId }) {
  useEffect(() => {
    const connection = createConnection(serverUrl, roomId);
    connection.connect();
    return () => connection.disconnect();
  }, [roomId]); // ✅ 所有声明的依赖项
  // ...
}
```

现在 `serverUrl` 不再是一个响应式值（并且在重新渲染时也不会更改），它就不需要成为一个依赖项。**如果 Effect 的代码不使用任何响应式值，则其依赖项列表应为空（`[]`）**：

```js {1,2,9}
const serverUrl = 'https://localhost:1234'; // 不再是响应式值
const roomId = 'music'; // 不再是响应式值

function ChatRoom() {
  useEffect(() => {
    const connection = createConnection(serverUrl, roomId);
    connection.connect();
    return () => connection.disconnect();
  }, []); // ✅ 所有声明的依赖项
  // ...
}
```

> [依赖项为空的 Effect](/learn/lifecycle-of-reactive-effects#what-an-effect-with-empty-dependencies-means) 不会在组件任何的 props 或 state 发生改变时重新运行。

Reviewer 3: ```suggestion
[依赖项为空数组的 Effect](/learn/lifecycle-of-reactive-effects#what-an-effect-with-empty-dependencies-means) 不会在组件任何的 props 或 state 发生改变时重新运行。
```

<Pitfall>

如果你有一个现有的代码库，可能会有一些 Effect 像这样抑制了代码检查工具：

```js {3-4}
useEffect(() => {
  // ...
  // 🔴 避免这样抑制代码检查工具：
  // eslint-ignore-next-line react-hooks/exhaustive-deps
}, []);
```

**当依赖项不匹配代码时，引入 bug 的风险很高**。通过抑制代码检查工具，你“欺骗”了 React 关于你 Effect 所依赖的值。[相反，证明它们是不必要的](/learn/removing-effect-dependencies#removing-unnecessary-dependencies)。

</Pitfall>

<Recipes titleText="传递响应式依赖项的示例" titleId="examples-dependencies">

#### 传递依赖项数组 {/*passing-a-dependency-array*/}

> 如果指定了依赖项，则 Effect 在 **初始渲染后 _以及_ 依赖项变更的重新渲染后** 运行。

Reviewer 3: ```suggestion
如果指定了依赖项，则 Effect 在 **初始渲染后以及依赖项变更的重新渲染后** 运行。
```

```js {3}
useEffect(() => {
  // ...
}, [a, b]); // 如果 a 或 b 不同则会再次运行
```

> 在下面的示例中，`serverUrl` 和 `roomId` 是 [响应式值](/learn/lifecycle-of-reactive-effects#effects-react-to-reactive-values)，因此它们都必须被指定为依赖项。因此，在下拉列表中选择不同的聊天室或编辑服务器 URL 输入框会导致重新连接聊天室。但是，由于 `message` 没有在 Effect 中使用（因此它不是依赖项），编辑消息不会重新连接聊天室。

Reviewer 3: ```suggestion
在下面的示例中，`serverUrl` 和 `roomId` 是 [响应式值](/learn/lifecycle-of-reactive-effects#effects-react-to-reactive-values)，因此它们都必须被指定为依赖项。这将导致在下拉列表中选择不同的聊天室或编辑服务器 URL 输入框会导致重新连接聊天室。但是，由于 `message` 没有在 Effect 中使用（因为它不是依赖项），编辑消息不会重新连接聊天室。
```

Reviewer 3: 抱歉，没有注意到后半句中出现了「导致」。翻译后的内容连续两个半句都出现了「因此」，可以考虑其他表达，避免重复表达

Author: 那我把第一个”因此“替换成：所以它们都必须被指定为依赖项

</Sandpack>

现在你已经在 Effect 内部创建了 `options` 对象，因此 Effect 仅依赖于 `roomId` 字符串。

通过此修复，在输入框中输入不会导致重新连接到聊天室。与会被重新创建的对象不同，像 `roomId` 这样的字符串只有在被设置为其它值时才会更改。[阅读有关删除依赖项的更多信息](/learn/removing-effect-dependencies)。

---

### 删除不必要的函数依赖项 {/*removing-unnecessary-function-dependencies*/}

> 如果你的 Effect 依赖于在渲染期间创建的对象或函数，则它可能会频繁运行。例如，此 Effect 在每次渲染后重新连接，因为 `createOptions` 函数[在每次渲染时都不同](/learn/removing-effect-dependencies#does-some-reactive-value-change-unintentionally)：

Reviewer 3: ```suggestion
如果你的 Effect 依赖于在渲染期间创建的对象或函数，则它可能会频繁运行。例如，此 Effect 在每次渲染后重新连接，因为 `createOptions` 函数 [在每次渲染时都不同](/learn/removing-effect-dependencies#does-some-reactive-value-change-unintentionally)：
```

</Sandpack>

现在你在 Effect 内部定义了 `createOptions` 函数，这样 Effect 本身只依赖于 `roomId` 字符串。通过此修复，输入框的输入不会重新连接聊天室。与被重新创建的函数不同，像 `roomId` 这样的字符串除非你将其设置为其它值，否则它不会改变。[了解更多有关移除依赖项的信息](/learn/removing-effect-dependencies)。 

---

### 从 Effect 读取最新的 props 和 state {/*reading-the-latest-props-and-state-from-an-effect*/}

<Wip>

本节描述了一个 **实验性的 API**，它还没有在一个稳定的 React 版本中发布。

</Wip>

默认情况下，在 Effect 中读取响应式值时，必须将其添加为依赖项。这样可以确保你的 Effect 对该值的每次更改都“作出响应”。对于大多数依赖项，这是你想要的行为。

> **然而，有时你想要从 Effect 中阅读 **最新的** props 和 state，而不“响应”它们**。例如，假设你想记录每次页面访问时购物车中的商品数量：

Reviewer 3: ```suggestion
**然而，有时你想要从 Effect 中获取 **最新的** props 和 state，而不“响应”它们**。例如，假设你想记录每次页面访问时购物车中的商品数量：
```

```js {3}
function Page({ url, shoppingCart }) {
  useEffect(() => {
    logVisit(url, shoppingCart.length);
  }, [url, shoppingCart]); // ✅ 所有声明的依赖项
  // ...
}
```

> **如果你想在每次 `url` 更改后记录一次新的页面访问，而不是在 `shoppingCart` 更改后记录，该怎么办**？你不能在不违反 [响应规则](#specifying-reactive-dependencies) 的情况下将 `shoppingCart` 从依赖项中移除。然而，你可以表达你 **不希望** 某些代码对更改做出“响应”，即使它是在 Effect 内部调用的。使用 [`useEffectEvent`](/reference/react/experimental_useEffectEvent) Hook [声明一个 **Effect 事件**](/learn/separating-events-from-effects#declaring-an-effect-event)，并将读取 `shoppingCart` 的代码移入其中：

Reviewer 3: ```suggestion
**如果你想在每次 `url` 更改后记录一次新的页面访问，而不是在 `shoppingCart` 更改后记录，该怎么办**？你不能在不违反 [响应规则](#specifying-reactive-dependencies) 的情况下将 `shoppingCart` 从依赖项中移除。然而，你可以表达你 **不希望** 某些代码对更改做出“响应”，即使它是在 Effect 内部调用的。使用 [`useEffectEvent`](/reference/react/experimental_useEffectEvent) Hook [声明 **Effect 事件**](/learn/separating-events-from-effects#declaring-an-effect-event)，并将读取 `shoppingCart` 的代码移入其中：
```

  useEffect(() => {
    onVisit(url);
  }, [url]); // ✅ 所有声明的依赖项
  // ...
}
```

**Effect 事件不是响应式的，必须始终省略其作为 Effect 的依赖项**。这就是让你在其中放置非响应式代码（可以在其中读取某些 props 和 state 的最新值）的原因。通过在 `onVisit` 中读取 `shoppingCart`，确保了 `shoppingCart` 不会使 Effect 重新运行。

[阅读更多关于 Effect Event 如何让你分离响应式和非响应式代码的内容](/learn/separating-events-from-effects#reading-latest-props-and-state-with-effect-events)。

---

### 在服务器和客户端上显示不同的内容 {/*displaying-different-content-on-the-server-and-the-client*/}

> 如果你的应用程序使用服务端（[直接](/reference/react-dom/server) 或通过 [框架](/learn/start-a-new-react-project#production-grade-react-frameworks)）渲染，你的组件将会在两个不同的环境中渲染。在服务器上，它将渲染生成初始 HTML。在客户端，React 将再次运行渲染代码，以便将事件处理附加到该 HTML 上。这就是为什么要让 [水合作用](/reference/react-dom/client/hydrateRoot#hydrating-server-rendered-html) 发挥作用，你的初始渲染输出必须在客户端和服务器上完全相同的原因。

Reviewer 3: ```suggestion
如果你的应用程序使用服务端（[直接](/reference/react-dom/server) 或通过 [框架](/learn/start-a-new-react-project#production-grade-react-frameworks)）渲染，你的组件将会在两个不同的环境中渲染。在服务器上，它将渲染生成初始 HTML。在客户端，React 将再次运行渲染代码，以便将事件处理附加到该 HTML 上。这就是为什么要让 [hydration](/reference/react-dom/client/hydrateRoot#hydrating-server-rendered-html) 发挥作用，你的初始渲染输出必须在客户端和服务器上完全相同的原因。
```

  }, []);

  if (didMount) {
    // ... 返回仅客户端的 JSX ...
  }  else {
    // ... 返回初始 JSX ...
  }
}
```

> 当应用加载时，用户首先会看到初始渲染的输出。然后，当它加载完进行水合时，你的 Effect 将会运行并且将 `didMount` 设置为 `true`，触发重新渲染。这将切换到仅在客户端的渲染输出。Effect 不在服务器上运行，这就是为什么 `didMount` 在初始服务器渲染期间为 `false` 的原因。

Reviewer 3: ```suggestion
当应用加载时，用户首先会看到初始渲染的输出。然后，当它加载完并进行水合（hydrated）时，Effect 将会运行并且将 `didMount` 设置为 `true`，从而触发重新渲染。这将切换到仅在客户端的渲染输出。Effect 不在服务器上运行，这就是为什么 `didMount` 在初始服务器渲染期间为 `false` 的原因。
```

谨慎使用此模式。请记住，网络连接速度较慢的用户将在相当长的时间内（可能是数秒钟）看到初始内容，因此你不希望对组件的外观进行突兀的更改。在许多情况下，你可以通过使用 CSS 条件性地显示不同的内容来避免这种需要。

---

## 疑难解答 {/*troubleshooting*/}

### Effect 在组件挂载时运行了两次 {/*my-effect-runs-twice-when-the-component-mounts*/}

> 当开启开发模式的 StrictMode 时，React 会在实际运行 setup 之前额外运行一次 setup 和 cleanup。

Reviewer 3: ```suggestion
在开发环境下，如果开启严格模式，React 会在实际运行 setup 之前额外运行一次 setup 和 cleanup。
```

> 这是一个压力测试，用于验证 Effect 的逻辑是否正确实现。如果出现可见问题，则 cleanup 函数缺少某些逻辑。cleanup 函数应该停止或撤消 setup 函数所做的任何操作。一般来说，用户不应该能够区分 setup 被调用一次（如在生产中）和调用 setup→cleanup→setup 序列（如在开发中）。

Reviewer 3: ```suggestion
这是一个压力测试，用于验证 Effect 的逻辑是否正确实现。如果出现可见问题，则 cleanup 函数缺少某些逻辑。cleanup 函数应该停止或撤消 setup 函数所做的任何操作。一般来说，用户不应该能够区分 setup 被调用一次（如在生产中）和调用 setup → cleanup → setup 序列（如在开发中）。
```

> 这是一个压力测试，用于验证 Effect 的逻辑是否正确实现。如果出现可见问题，则 cleanup 函数缺少某些逻辑。cleanup 函数应该停止或撤消 setup 函数所做的任何操作。一般来说，用户不应该能够区分 setup 被调用一次（如在生产中）和调用 setup → cleanup → setup 序列（如在开发中）。

Reviewer 3: ```suggestion
这是一个压力测试，用于验证 Effect 的逻辑是否正确实现。如果出现可见问题，则 cleanup 函数缺少某些逻辑。cleanup 函数应该停止或撤消 setup 函数所做的任何操作。一般来说，用户不应该能够区分 setup 被调用一次（如在生产环境中）和调用 setup → cleanup → setup 序列（如在开发环境中）。
```

  }, [serverUrl, roomId]);
```

[了解 Effect 生命周期与组件的生命周期有何不同](/learn/lifecycle-of-reactive-effects#the-lifecycle-of-an-effect)。

---

### 我的 Effect 做了一些视觉相关的事情，在它运行之前我看到了一个闪烁 {/*my-effect-does-something-visual-and-i-see-a-flicker-before-it-runs*/}

> 如果你的 Effect 一定要阻止浏览器绘制屏幕，使用 [`useLayoutEffect`](/reference/react/useLayoutEffect) 替换 `useEffect`。请注意，**绝大多数的 Effect 都不需要这样**。只有当在浏览器绘制之前运行 Effect 非常重要的时候才需要如此：例如，在用户看到 tootip 之前测量和定位它。

Reviewer 3: ```suggestion
如果 Effect 一定要阻止浏览器绘制屏幕，使用 [`useLayoutEffect`](/reference/react/useLayoutEffect) 替换 `useEffect`。请注意，**绝大多数的 Effect 都不需要这样**。只有当在浏览器绘制之前运行 Effect 非常重要的时候才需要如此：例如，在用户看到 tooltip 之前测量并定位它。
```
